File: shared/nutrition-engine/calculator.ts
/**
 * منفذ من nutrition_ai/calculator.py — نقطة الحقيقة الوحيدة للأرقام. الجزء المستقل عن DB
 * (compute_food/totals_for_items/day_utc_range/today_utc_range) + الجزء المعتمد على Repository
 * (today_totals/meals_by_type_for_day/today_water_ml)، بعد اكتمال طبقة db/repository.ts.
 */
import * as foodSearch from "./foodSearch.js";
import * as iraqTime from "./iraqTime.js";
import type { Repository, MealLogRecord } from "./db/repository.js";

export async function computeFood(foodId: number, grams: number) {
  return foodSearch.computeNutrition(foodId, grams);
}

export interface NutritionItem {
  calories: number;
  protein: number;
  carbs: number;
  fat: number;
}

export function totalsForItems(items: NutritionItem[]) {
  return {
    calories: items.reduce((sum, i) => sum + i.calories, 0),
    protein: items.reduce((sum, i) => sum + i.protein, 0),
    carbs: items.reduce((sum, i) => sum + i.carbs, 0),
    fat: items.reduce((sum, i) => sum + i.fat, 0),
  };
}

/**
 * نفس منطق today_utc_range لكن ليوم بغدادي محدد (بدل اليوم الحالي إجباريًا). يرجّع حدود UTC
 * [بداية، نهاية) مطابقة لتخزين created_at بـUTC ساذج (توافقًا مع نمط التخزين الحالي بايثون).
 */
export function dayUtcRange(targetYear: number, targetMonth: number, targetDay: number): [Date, Date] {
  // بغداد UTC+3 ثابت بدون توقيت صيفي — منتصف ليل بغداد لليوم المطلوب = 21:00 UTC لليوم السابق
  const startUtc = new Date(Date.UTC(targetYear, targetMonth - 1, targetDay, -3, 0, 0));
  const endUtc = new Date(startUtc.getTime() + 24 * 60 * 60 * 1000);
  return [startUtc, endUtc];
}

/** يرجّع حدود UTC لليوم الحالي بتوقيت بغداد. */
export function todayUtcRange(now: Date = new Date()): [Date, Date] {
  const { year, month, day } = iraqTime.nowBaghdad(now);
  return dayUtcRange(year, month, day);
}

export interface DayTotals {
  calories: number;
  protein: number;
  carbs: number;
  fat: number;
  logs: MealLogRecord[];
}

export async function todayTotals(repo: Repository, userId: string, now: Date = new Date()): Promise<DayTotals> {
  const [start, end] = todayUtcRange(now);
  const logs = await repo.findMealLogsInRange(userId, start, end);
  return {
    calories: logs.reduce((s, l) => s + l.total_calories, 0),
    protein: logs.reduce((s, l) => s + l.total_protein, 0),
    carbs: logs.reduce((s, l) => s + l.total_carbs, 0),
    fat: logs.reduce((s, l) => s + l.total_fat, 0),
    logs,
  };
}

export async function todayWaterMl(repo: Repository, userId: string, now: Date = new Date()): Promise<number> {
  const [start, end] = todayUtcRange(now);
  const rows = await repo.findWaterLogsInRange(userId, start, end);
  return rows.reduce((s, r) => s + r.ml, 0);
}

export type MealBucket =
  | { status: "NOT_STARTED" }
  | {
      status: "LOGGED"; calories: number; protein: number; carbs: number; fat: number;
      foods: string[]; logged_at: Date;
    };

export interface MealsByType {
  breakfast: MealBucket;
  lunch: MealBucket;
  dinner: MealBucket;
  snack: MealBucket[];
}

/**
 * يجمّع وجبات يوم بغدادي محدد حسب نوعها — breakfast/lunch/dinner كل وحدة صف واحد (أو
 * NOT_STARTED لو ماكو)، snack قائمة (يدعم أكثر من سناك بنفس اليوم).
 */
export async function mealsByTypeForDay(
  repo: Repository, userId: string, targetYear: number, targetMonth: number, targetDay: number,
): Promise<MealsByType> {
  const [start, end] = dayUtcRange(targetYear, targetMonth, targetDay);
  const logs = (await repo.findMealLogsInRange(userId, start, end))
    .sort((a, b) => a.created_at.getTime() - b.created_at.getTime());

  const summarize = (log: MealLogRecord): MealBucket => {
    let foods: string[] = [];
    try {
      foods = log.matched_foods_json ? JSON.parse(log.matched_foods_json) : [];
    } catch {
      foods = [];
    }
    return {
      status: "LOGGED", calories: log.total_calories, protein: log.total_protein,
      carbs: log.total_carbs, fat: log.total_fat, foods, logged_at: log.created_at,
    };
  };

  const result: MealsByType = {
    breakfast: { status: "NOT_STARTED" }, lunch: { status: "NOT_STARTED" },
    dinner: { status: "NOT_STARTED" }, snack: [],
  };
  for (const log of logs) {
    if (log.meal_type === "snack") {
      result.snack.push(summarize(log));
    } else if (log.meal_type === "breakfast" || log.meal_type === "lunch" || log.meal_type === "dinner") {
      result[log.meal_type] = summarize(log);
    }
  }
  return result;
}

File: nutrition_ai/calculator.py
"""
Nutrition Calculator — نقطة الحقيقة الوحيدة للأرقام. أي مكان يحتاج سعرات/ماكروز يمرّ من هنا،
وهذا بدوره يمرّ حصرًا من food_search.py (قاعدة foods.sqlite). لا AI يحسب أي رقم بنفسه.
"""
import json
from datetime import datetime, time, timedelta, timezone

import iraq_time
import food_search


def compute_food(food_id: int, grams: float) -> dict:
    return food_search.compute_nutrition(food_id, grams)


def totals_for_items(items: list) -> dict:
    return {
        "calories": sum(i["calories"] for i in items),
        "protein": sum(i["protein"] for i in items),
        "carbs": sum(i["carbs"] for i in items),
        "fat": sum(i["fat"] for i in items),
    }


def day_utc_range(target_date) -> tuple[datetime, datetime]:
    """نفس منطق today_utc_range لكن ليوم بغدادي محدد (date) بدل اليوم الحالي إجباريًا —
    يُستخدم لعرض أيام سابقة بصفحة 'يومي الغذائي' بدون تكرار منطق التحويل الزمني."""
    start_local = datetime.combine(target_date, time.min, tzinfo=iraq_time.BAGHDAD_TZ)
    end_local = start_local + timedelta(days=1)
    start_utc = start_local.astimezone(timezone.utc).replace(tzinfo=None)
    end_utc = end_local.astimezone(timezone.utc).replace(tzinfo=None)
    return start_utc, end_utc


def today_utc_range() -> tuple[datetime, datetime]:
    """
    يرجّع (بداية، نهاية) اليوم الحالي بتوقيت بغداد، محوّلة لنطاق UTC ساذج (بدون tzinfo) —
    لازم نقارن بيه لأن created_at يُخزَّن عبر datetime.utcnow(). مقارنة date.today() (توقيت
    الجهاز المحلي) مباشرة مع تاريخ عمود UTC كانت تسبب إسقاط وجبات اليوم قرب منتصف الليل
    (بغداد UTC+3) — أي وجبة بين 21:00-23:59 UTC تحسب "أمس" محليًا بالخطأ.
    """
    return day_utc_range(iraq_time.now_baghdad().date())


def today_totals(user, MealLog):
    start_utc, end_utc = today_utc_range()
    logs = MealLog.query.filter(
        MealLog.user_id == user.id,
        MealLog.created_at >= start_utc,
        MealLog.created_at < end_utc,
    ).all()
    total_cal = sum(l.total_calories for l in logs)
    total_protein = sum(l.total_protein for l in logs)
    total_carbs = sum(l.total_carbs for l in logs)
    total_fat = sum(l.total_fat for l in logs)
    return {
        "calories": total_cal, "protein": total_protein,
        "carbs": total_carbs, "fat": total_fat, "logs": logs,
    }


def meals_by_type_for_day(user, MealLog, target_date) -> dict:
    """يجمّع وجبات يوم بغدادي محدد حسب نوعها — من نفس صفوف MealLog الحقيقية (created_at)،
    بدون أي حساب سعرات جديد (المجاميع مأخوذة من الأعمدة المحسوبة مسبقًا بـ_finalize_meal).
    breakfast/lunch/dinner كل وحدة صف واحد (أو None لو ماكو)، snack قائمة (يدعم أكثر من سناك
    بنفس اليوم)."""
    start_utc, end_utc = day_utc_range(target_date)
    logs = MealLog.query.filter(
        MealLog.user_id == user.id,
        MealLog.created_at >= start_utc,
        MealLog.created_at < end_utc,
    ).order_by(MealLog.created_at.asc()).all()

    def _summarize(log):
        try:
            items = json.loads(log.matched_foods_json) if log.matched_foods_json else []
        except (ValueError, TypeError):
            items = []
        return {
            "status": "LOGGED",
            "calories": log.total_calories, "protein": log.total_protein,
            "carbs": log.total_carbs, "fat": log.total_fat,
            "foods": items, "logged_at": log.created_at,
        }

    result = {"breakfast": None, "lunch": None, "dinner": None, "snack": []}
    for log in logs:
        if log.meal_type == "snack":
            result["snack"].append(_summarize(log))
        elif log.meal_type in result:
            result[log.meal_type] = _summarize(log)

    for meal_type in ("breakfast", "lunch", "dinner"):
        if result[meal_type] is None:
            result[meal_type] = {"status": "NOT_STARTED"}

    return result


def today_water_ml(user, WaterLog) -> int:
    start_utc, end_utc = today_utc_range()
    rows = WaterLog.query.filter(
        WaterLog.user_id == user.id,
        WaterLog.created_at >= start_utc,
        WaterLog.created_at < end_utc,
    ).all()
    return sum(r.ml for r in rows)
